File: chat-uikit/src/main/java/io/agora/chat/uikit/chathistory/EaseChatHistoryLayout.java
package io.agora.chat.uikit.chathistory;

import android.content.Context;
import android.util.AttributeSet;
import android.view.LayoutInflater;
import android.widget.RelativeLayout;

import androidx.appcompat.app.AppCompatActivity;

import java.util.List;

import io.agora.chat.ChatMessage;
import io.agora.chat.uikit.R;
import io.agora.chat.uikit.chat.adapter.EaseMessageAdapter;
import io.agora.chat.uikit.chat.widget.EaseChatMessageListLayout;
import io.agora.chat.uikit.chathistory.presenter.EaseChatHistoryPresenter;
import io.agora.chat.uikit.chathistory.presenter.EaseChatHistoryPresenterImpl;
import io.agora.chat.uikit.chathistory.presenter.IChatHistoryLayoutView;
import io.agora.chat.uikit.chathistory.presenter.IHandleChatHistory;

public class EaseChatHistoryLayout extends RelativeLayout implements IChatHistoryLayoutView, IHandleChatHistory {
    private EaseChatMessageListLayout messageListLayout;
    private EaseChatHistoryPresenter mPresenter;
    private Context mContext;
    private EaseMessageAdapter messageAdapter;

    public EaseChatHistoryLayout(Context context) {
        this(context, null);
    }

    public EaseChatHistoryLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public EaseChatHistoryLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mContext = context;
        LayoutInflater.from(context).inflate(R.layout.ease_layout_chat_history, this);
        initView();
        initListener();
        initData();
    }

    private void initView() {
        messageListLayout = findViewById(R.id.layout_chat_message);
        messageListLayout.canUseDefaultRefresh(false);
    }

    private void initListener() {

    }

    private void initData() {
        mPresenter = new EaseChatHistoryPresenterImpl();
        if(mContext instanceof AppCompatActivity) {
            ((AppCompatActivity) mContext).getLifecycle().addObserver(mPresenter);
        }
        mPresenter.attachView(this);
        messageAdapter = new EaseChatHistoryAdapter();
        messageListLayout.setMessageAdapter(messageAdapter);
    }

    public void loadData(ChatMessage message) {
        if(message == null) {
            finishCurrentChat();
            return;
        }
        if(message.getType() != ChatMessage.Type.COMBINE) {
            finishCurrentChat();
            return;
        }
        mPresenter.downloadCombineMessage(message);
    }

    protected void finishCurrentChat() {

    }

    @Override
    public void downloadCombinedMessagesSuccess(List<ChatMessage> messageList) {
        messageListLayout.setData(messageList);
    }

    @Override
    public void downloadCombinedMessagesFailed(int error, String errorMsg) {

    }

    @Override
    public void downloadThumbnailSuccess(ChatMessage message, int position) {
        messageListLayout.refreshMessage(message);
    }

    @Override
    public void downloadThumbnailFailed(ChatMessage message, int position, int error, String errorMsg) {

    }

    @Override
    public void downloadVoiceSuccess(ChatMessage message, int position) {
        messageListLayout.refreshMessage(message);
    }

    @Override
    public void downloadVoiceFailed(ChatMessage message, int position, int error, String errorMsg) {

    }

    @Override
    public void refreshAll() {
        messageListLayout.refreshMessages();
    }

    @Override
    public void refreshItem(ChatMessage message, int position) {
        messageListLayout.refreshMessage(message);
    }

    @Override
    public Context context() {
        return mContext;
    }

    @Override
    public void setPresenter(EaseChatHistoryPresenter presenter) {
        this.mPresenter = presenter;
        if(mContext instanceof AppCompatActivity) {
            ((AppCompatActivity) mContext).getLifecycle().addObserver(mPresenter);
        }
        mPresenter.attachView(this);
    }

    @Override
    public void setMessageAdapter(EaseMessageAdapter adapter) {
        if(adapter != null) {
            this.messageAdapter = adapter;
            messageListLayout.setMessageAdapter(messageAdapter);
        }
    }

    @Override
    public EaseChatMessageListLayout getChatMessageListLayout() {
        return messageListLayout;
    }
}

File: ease-im-kit/src/main/kotlin/io/agora/uikit/feature/chat/chathistory/EaseChatHistoryLayout.kt
package io.agora.uikit.feature.chat.chathistory

import android.content.Context
import android.util.AttributeSet
import android.view.LayoutInflater
import android.widget.RelativeLayout
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import io.agora.uikit.common.ChatMessage
import io.agora.uikit.common.ChatMessageType
import io.agora.uikit.common.extensions.mainScope
import io.agora.uikit.databinding.EaseLayoutChatHistoryBinding
import io.agora.uikit.feature.chat.adapter.EaseMessagesAdapter
import io.agora.uikit.feature.chat.config.EaseChatMessageItemConfig
import io.agora.uikit.feature.chat.interfaces.IChatHistoryLayout
import io.agora.uikit.feature.chat.interfaces.IChatHistoryResultView
import io.agora.uikit.feature.chat.interfaces.OnCombineMessageDownloadCallback
import io.agora.uikit.feature.chat.widgets.EaseChatMessageListLayout
import io.agora.uikit.viewmodel.chathistory.EaseChatHistoryViewModel
import io.agora.uikit.viewmodel.chathistory.IChatHistoryRequest
import kotlinx.coroutines.launch

open class EaseChatHistoryLayout @JvmOverloads constructor(
    private val context: Context,
    private val attrs: AttributeSet? = null,
    private val defStyleAttr: Int = 0
): RelativeLayout(context, attrs, defStyleAttr), IChatHistoryLayout, IChatHistoryResultView {

    /**
     * The view model of the chat history view.
     */
    private var viewModel: IChatHistoryRequest? = null

    /**
     * The adapter of the chat history view.
     */
    private var messagesAdapter: EaseMessagesAdapter? = null

    /**
     * The callback of the combine message download or parse.
     */
    private var historyCallback: OnCombineMessageDownloadCallback? = null

    protected val binding: EaseLayoutChatHistoryBinding by lazy { EaseLayoutChatHistoryBinding.inflate(
        LayoutInflater.from(context), this, true) }

    init {
        initView()
    }

    private fun initView() {
        viewModel = if (context is AppCompatActivity) {
            ViewModelProvider(context)[EaseChatHistoryViewModel::class.java]
        } else {
            EaseChatHistoryViewModel()
        }
        viewModel?.attachView(this)

        // Set the history adapter for the EaseChatMessageListLayout
        messagesAdapter = EaseChatHistoryAdapter()
        messagesAdapter?.setItemConfig(EaseChatMessageItemConfig(context, null))
        binding.root.setMessagesAdapter(messagesAdapter)

        binding.root.refreshLayout?.let {
            it.setEnableLoadMore(false)
            it.setEnableRefresh(false)
        }
    }

    fun loadData(message: ChatMessage?) {
        if (message == null) {
            return
        }
        if (message.type != ChatMessageType.COMBINE) {
            return
        }
        viewModel?.downloadCombineMessage(message)
    }

    override fun setViewModel(viewModel: IChatHistoryRequest?) {
        this.viewModel = viewModel
        this.viewModel?.attachView(this)
    }

    override fun setMessagesAdapter(adapter: EaseMessagesAdapter?) {
        adapter?.let {
            this.messagesAdapter = adapter
            binding.root.setMessagesAdapter(messagesAdapter)
        }
    }

    override fun setOnCombineMessageDownloadCallback(callback: OnCombineMessageDownloadCallback?) {
        this.historyCallback = callback
    }

    override fun getChatMessageListLayout(): EaseChatMessageListLayout {
        return binding.root
    }

    override fun downloadCombinedMessagesSuccess(messageList: List<ChatMessage>) {
        context.mainScope().launch {
            messagesAdapter?.setData(messageList.toMutableList())
            historyCallback?.onDownloadSuccess(messageList)
        }

    }

    override fun downloadCombinedMessagesFail(error: Int, errorMsg: String?) {
        context.mainScope().launch {
            historyCallback?.onDownloadFail(error, errorMsg)
        }
    }


}